lib/mayaUsd/undo/OpUndoInfo.h

```diff
@@ -0,0 +1,148 @@
+//
+#ifndef MAYAUSD_UNDO_OP_UNDO_INFO_H
+#define MAYAUSD_UNDO_OP_UNDO_INFO_H
+
+#include <mayaUsd/base/api.h>
+
+#include <maya/MObjectHandle.h>
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <unordered_set>
+#include <vector>
+
+namespace MAYAUSD_NS_DEF {
+
+//------------------------------------------------------------------------------
+// OpUndoItem
+//------------------------------------------------------------------------------
+
+/// \class OpUndoItem
+/// \brief Record data needed to undo or redo a single undo sub-operation.
+///
+/// See OpUndoItems.h for concrete implementations.
+
+class OpUndoItem
+{
+public:
+    typedef std::unique_ptr<OpUndoItem> Ptr;
+
+    /// \brief construct a single named sub-operation.
+    MAYAUSD_CORE_PUBLIC
+    OpUndoItem(std::string name)
+        : _name(std::move(name))
+    {
+    }
+
+    MAYAUSD_CORE_PUBLIC
+    virtual ~OpUndoItem() = default;
+
+    /// \brief undo a single sub-operation.
+    MAYAUSD_CORE_PUBLIC
+    virtual bool undo() = 0;
+
+    /// \brief redo a single sub-operation.
+    MAYAUSD_CORE_PUBLIC
+    virtual bool redo() = 0;
+
+    /// \brief get the undo item name, used for debugging and logging.
+    MAYAUSD_CORE_PUBLIC
+    const std::string& getName() const { return _name; }
+
+private:
+    // Name the undo items to help debugging, tracing and logging.
+    std::string _name;
+};
+
+//------------------------------------------------------------------------------
+// OpUndoInfo
+//------------------------------------------------------------------------------
+
+/// \class OpUndoInfo
+/// \brief Record everything needed to undo or redo a complete operation or command.
+
+class OpUndoInfo
+{
+public:
+    /// \brief construct an undo info.
+    MAYAUSD_CORE_PUBLIC
+    OpUndoInfo() = default;
+
+    MAYAUSD_CORE_PUBLIC
+    OpUndoInfo(OpUndoInfo&&) = default;
+
+    MAYAUSD_CORE_PUBLIC
+    OpUndoInfo& operator=(OpUndoInfo&&) = default;
+
+    /// \brief destroy the undo info.
+    MAYAUSD_CORE_PUBLIC
+    ~OpUndoInfo();
+
+    /// \brief undo a complete PrimUpdaterManager operation.
+    MAYAUSD_CORE_PUBLIC
+    bool undo();
+
+    /// \brief redo a complete PrimUpdaterManager operation.
+    MAYAUSD_CORE_PUBLIC
+    bool redo();
+
+    /// \brief add a undo item. This takes ownership of the item.
+    MAYAUSD_CORE_PUBLIC
+    void addItem(OpUndoItem::Ptr&& item);
+
+    /// \brief add an object that was deleted, to help avoid deleting an object twice.
+    MAYAUSD_CORE_PUBLIC
+    void addDeleted(const MObjectHandle obj);
+
+    /// \brief verify if an object was already deleted, to help avoid deleting an object twice.
+    MAYAUSD_CORE_PUBLIC
+    bool isDeleted(const MObjectHandle obj) const;
+
+    /// \brief clear all undo/redo information contained here.
+    MAYAUSD_CORE_PUBLIC
+    void clear();
+
+    /// \brief extract all undo/redo information contained here and clear.
+    MAYAUSD_CORE_PUBLIC
+    OpUndoInfo extract();
```

Comment:
Well, I did not want to leave the OpUndoInfo object in the invalid state that a move assignment / constructor would.  This is especially important because there is a permanent instance in the UsdUndoManager singleton.